neqo-transport/src/connection/tests/stream.rs

```diff
@@ -111,6 +113,121 @@ fn transfer() {
     assert!(fin3);
 }
 
+#[cfg(test)]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct IdEntry {
+    sendorder: StreamOrder,
+    stream_id: StreamId,
+}
+
+#[cfg(test)]
+// tests stream sendorder priorization
+fn sendorder_test(order_of_sendorder: &Vec<StreamOrder>) {
+    let mut client = default_client();
+    let mut server = default_server();
+    connect_force_idle(&mut client, &mut server);
+
+    qdebug!("---- client sends");
+    // open all streams and set the sendorders
+    let mut ordered = BinaryHeap::new();
+    let mut streams = Vec::<StreamId>::new();
+    for sendorder in order_of_sendorder {
+	streams.push(client.stream_create(StreamType::UniDi).unwrap());
+	let id = *(streams.last()).unwrap();
+	ordered.push(IdEntry { sendorder: *sendorder, stream_id: id });
+	client.streams.set_sendorder(id, sendorder.sendorder).ok();
+    }
+    // Write some data to all the streams
+    for stream_id in streams {
+	client.stream_send(stream_id, &[6; 100]).unwrap();
+    }
+
+    // Sending this much takes a few datagrams.
+    let mut datagrams = vec![];
+    let mut out = client.process_output(now());
+    while let Some(d) = out.dgram() {
+        datagrams.push(d);
+        out = client.process_output(now());
+    }
+    assert_eq!(*client.state(), State::Confirmed);
+
+    qdebug!("---- server receives");
+    for (d_num, d) in datagrams.into_iter().enumerate() {
+        let out = server.process(Some(d), now());
+        assert_eq!(
+            out.as_dgram_ref().is_some(),
+            (d_num + 1) % usize::try_from(DEFAULT_ACK_PACKET_TOLERANCE + 1).unwrap() == 0
+        );
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+    }
+    assert_eq!(*server.state(), State::Confirmed);
+
+    let mut stream_ids = server.events().filter_map(|evt| match evt {
+        ConnectionEvent::RecvStreamReadable { stream_id, .. } => Some(stream_id),
+        _ => None,
+    });
+
+    // streams should arrive in priority order, not order of creation, if sendorder prioritization
+    // is working correctly
+    let mut recv_streams = Vec::<StreamId>::new();
+    let mut should_recv_streams =  Vec::<StreamId>::new();
+    // walk the ordered list of streams, and handle random ordering within a sendorder
+    let mut last_sendorder: StreamOrder = StreamOrder{sendorder: None};
+    // BinaryHeap's sorted iterator is nightly only for now
+    // the ordering is inverted, however!
+    let mut ordered_vec = ordered.into_sorted_vec();
+    ordered_vec.reverse();
+    for identry in ordered_vec.iter() {
```

Comment:
There might be an easier way:

Collection 1 is the declared send order, taken from the input list.  It consists of a Vec containing tuples of `(StreamId, Option<SendOrder>)`.  (Your IdEntry, though I wouldn't bother declaring a type for it.)

Collection 2 is the arrival order.  It consists of a map from `StreamId` to a simple counter that is incremented for each piece of data that arrives (without overwriting as some data might arrive twice).  (There is a clever way of constructing this.  You take your filter_map call above and then add `.enumerate().map(|(a, b)| (b, a)).collect::<HashMap<_, _>>()`)

The first collection is sorted by the SendOrder, but if two items have equal SendOrder, you use the arrival order from the second list to determine order (use `Vec::sort_unstable_by`).

Then you can check that the order of the stream identifiers in the first and second collections are equal.

Reply:
Perhaps this is better; it would be shorter.   I also think (though perhaps it's just because I'm new to Rust) that this will be harder to understand or modify, if I understand correctly.
I've been trying to code this up, and I can't figure out how to do "The first collection is sorted by the SendOrder, but if two items have equal SendOrder, you use the arrival order from the second list to determine order (use Vec::sort_unstable_by)."
I see https://doc.rust-lang.org/std/vec/struct.Vec.html#method.sort_unstable_by, but I don't see how to go  from that to a working fancy sorter.